docs/learn.md

```diff
@@ -721,20 +723,1250 @@ template](https://github.com/melange-re/melange-opam-template).
+indices to the `bs.as` decorator:
+
+```ocaml
+type t = {
+  foo : int; [@bs.as "0"]
+  bar : string; [@bs.as "1"]
+}
+
+let value = { foo = 7; bar = "baz" }
+```
+
+And its JavaScript generated code:
+
+```js
+var value = [
+  7,
+  "baz"
+];
+```
+
+##### Using `Js.t` objects
+
+Alternatively to records, Melange offers another type that can be used to
+produce JavaScript objects. This type is `'a Js.t`, where `'a` is an [OCaml
+object](https://v2.ocaml.org/manual/objectexamples.html).
+
+The advantage of objects versus records is that no type declaration is needed in
+advanced, which can be helpful for prototyping or quickly generating JavaScript
+object literals.
+
+Melange provides some ways to create `Js.t` object values, as well as accessing
+the properties inside them. To create values, the `[%bs.obj]` extension is used,
+and the `##` infix operator allows to read from the object properties:
+
+```ocaml
+let john = [%bs.obj { name = "john"; age = 99 }]
+let t = john##name
+```
+
+Which generates:
+
+```js
+var john = {
+  name: "john",
+  age: 99
+};
+
+var t = john.name;
+```
+
+Note that type inference with objects is more lenient than with records. For
+example, one can create a function that operates in all the object types that
+include a field `name` that is of type string, e.g.:
+
+```ocaml
+let name_extended obj = obj##name ^ " wayne"
+
+let one = name_extended [%bs.obj { name = "john"; age = 99 }]
+let two = name_extended [%bs.obj { name = "jane"; address = "1 infinite loop" }]
+```
+
+This is possible thanks to [objects
+polymorphism](https://v2.ocaml.org/manual/objectexamples.html#s%3Apolymorphic-methods),
+and is a characteristic shared by some types in OCaml, like objects and
+polymorphic variants. The book Real World OCaml has [a good
+introduction](https://dev.realworldocaml.org/objects.html) to object
+polymorphism.
+
+##### Bind to object properties
+
+If you need to bind only to the property of a JavaScript object, you can use
+`bs.get` and `bs.set`:
+
+```ocaml
+type textarea
+external set_name : textarea -> string -> unit = "name" [@@bs.set]
+external get_name : textarea -> string = "name" [@@bs.get]
+```
+
+You can also use `bs.get_index` and `bs.set_index` to access a dynamic property
+or an index:
+
+```ocaml
+type t
+external create : int -> t = "Int32Array" [@@bs.new]
+external get : t -> int -> int = "get" [@@bs.get_index]
+external set : t -> int -> int -> unit = "set" [@@bs.set_index]
+
+let () =
+  let i32arr = (create 3) in
+  set i32arr 0 42;
+  Js.log (get i32arr 0)
+```
+
+Which generates:
+
+```js
+var i32arr = new Int32Array(3);
+i32arr[0] = 42;
+console.log(i32arr[0]);
+```
+
+#### Objects with dynamic shape (dictionary-like)
+
+If you are binding to a JavaScript object that:
+
+- might or might not add or remove keys
+- contains only values that are of the same type
+
+Then it’s not really an object, it’s a dictionary or map. Use
```

Comment:
let's say "then you can model it in OCaml using `Js.Dict`."

> "it's not really an object"

but, it is :P

Reply:
Rewrote this section fully in https://github.com/melange-re/documentation-site/pull/12/commits/019bf1a6c0d9429bdb97737f375f39471183cc80, please lmk what you think.

Reply:
very nice!